계산기 [STEP 1] Whales

WhalesJin opened the pull request:
안녕하세요 원비(@wonbi92)🤵🏻‍♂️👰🏻‍♀️
드디어드디어 원비의 리뷰를 받게된 웨일🐋입니다!!!
계산기 프로젝트(부제: 고래와 이꾸라데쓰까?🧮) 잘 부탁드립니다아🙇🏻‍♀️

### 📝 고민한 점
> `Queue` 타입 구현
- `Queue` 타입을 구현을 위한 List 타입 선택 과정에 고민을 많이 했습니다.
그 중 두 가지를 가지고 비교를 해보았습니다.
    - 배열
        <img src = "https://github.com/yagom-academy/ios-calculator-app/assets/124643545/b6a3f94e-3d3f-4637-b6ef-9059f344ed73" width = "300"/>
        - 여러가지 데이터를 하나의 줄 모양으로 저장
        - 크기가 정해져 있다.
        - 메모리를 항상 차지하고 있다는 단점이 있다.
        - 크기가 똑같은 걸 아니까 100번째가 궁금하면 점프할 수 있다. -> 속도가 빠르다.
        (index를 통해)
        - 새로운 요소 삽입이 빠르다.
    - 단방향 연결리스트
        <img src = "https://github.com/yagom-academy/ios-calculator-app/assets/124643545/3679b5f6-e7ca-42a2-a32d-0579923fc358" width = "400"/>
        - 각각 떨어진 공간에 존재하는 데이터를 연결해 놓은 것
        - 실질적인 부분과 다음 주소값의 정보를 가지는 부분으로 이루어져있다.
        - 한 방향만 보고 있기 때문에 다음만 안다.
        - 주소값 정보만 바꿔주면 되어서 추가 / 삭제가 용이하다.
        - 추가/삭제가 자유로워 메모리 할당이 안된다. 한 번에 찾을 수 없다.
        - 저장 공간 효율이 높지 않다.
- 단방향 연결리스트의 가장 장점은 리스트 중간에서의 추가 / 삭제가 자유롭다는 부분인데 계산기 프로젝트에서는 중간 수정이 없어서 단방향 연결리스트보다는 크기가 정해져있고 새로운 요소 삽입이 빠른 배열이 더 효율적이라고 생각해서 배열로 선택하였습니다.

> `CalculatorItemQueue`의 메서드
- `Queue`을 구현할 때 있어야 하는 메서드들, 큐에 넣어주는 `enqueue()` 와 큐에 넣은 걸 다시 삭제하는 `dequeue()`를 구현했습니다.
- 추가적으로 계산기를 실행하다보면 초기화하는 경우도 있어서 데이터들을 다 삭제하는 `clearQueue()`라는 메서드도 구현해주었습니다.
- 또한, `test`를 진행하던 중 배열을 가지고 직접 비교해야하는 경우가 있어서 은닉화되어있는 배열을 볼 수 있도록 `readCalculatorItemQueue()`라는 함수도 구현하였습니다.
<br/>

### 🔍 조언을 얻고 싶은 부분

> `clearQueue()` 메서드
- `AC`버튼을 누르면 초기화 되어야해서 `clearQueue()`라는 메서드를 만들었는데, 
인스턴스 자체가 초기화가 된다면 큐 타입 안에 따로 메서드를 만들 필요가 없지 않을까 라는 생각이 들었습니다.
- 두 가지 생각에서 고민하다가 메서드를 만들었는데 `인스턴스 자체`를 초기화 하는 쪽과 메서드로 큐 안에 있는 `데이터만` 초기화 하는 쪽 두 가지 중 어떤 게 더 좋은 방법일지 조언을 얻고 싶습니다.

> `TDD` 시도와 `commit`
- `TDD`를 처음으로 진행해봤는데요, 너무 낯설고 이상해서 제대로 하고 있는 게 맞는지 의문이 들었습니다.
- `test`를 먼저 만들면 함수가 아직 만들어지지 않은 상태이기에 에러가 나있는 상태로 커밋을 하게 됩니다. 에러가 있는데 `commit`을 한다는게 너무 이상한데 `TDD`의 과정을 남기기 위해서 일단 그 상태로 `commit`을 진행했습니다.
이 `commit` 방법이 맞는지 더 좋은 다른 방법이 있는지 조언 부탁드립니다.🙏🏻
- 또 `Unit test`를 진행할 때는 테스트 함수 명을 `test_함수이름`으로 시작하게 짓는다고 들었는데, 그러면 함수 이름을 먼저 만들어두면 좋은지, 아니면 테스트 만들고`commit` 후에 함수 만들면서 테스트 함수 이름도 수정해야하는지 혼란스럽습니다. 좋은 조언이 있으시다면 부탁드립니다 🙏🏻

WhalesJin committed:
feat: 단위 테스트를 위한 타깃 생성

WhalesJin committed:
docs: 클래스 다이어그램

- 기본 틀만 구현

WhalesJin committed:
feat: CalculatorItem 프로토콜 생성

WhalesJin committed:
feat: CalculatorItemQueue 구조체 생성

요소는 CalculatorItem 프로토콜 준수

WhalesJin committed:
test: 원소를 추가하는 enqueue 함수의 테스트 함수 생성

WhalesJin committed:
test: 원소를 추가하는 enqueue 함수의 테스트 함수 수정

WhalesJin committed:
feat: 원소를 추가하는 enqueue 함수 생성

test 성공

WhalesJin committed:
test: 원소를 추가하는 enqueue 함수의 테스트 함수

- 원소가 배열에 제대로 들어가는지 확인(확인을 위해 일시적으로 은닉화 해제)

WhalesJin committed:
test: 배열의 첫번째 원소를 삭제하는 dequeue 함수의 테스트 함수

- 배열의 첫번째 원소가 삭제되는지 확인(확인을 위해 일시적으로 은닉화 해제)

WhalesJin committed:
feat: 배열의 첫번째 원소를 삭제하는 dequeue 함수 생성

test 성공

WhalesJin committed:
test: 배열의 모든 원소를 삭제하는 함수의 테스트 함수

확인을 위해 일시적으로 은닉화 해제

WhalesJin committed:
feat: 배열의 모든 원소를 삭제하는 clearQueue 함수 생성

test 성공

WhalesJin committed:
feat: calculatorItemQueue 읽어오는 함수 생성

- calculatorItemQueue 배열 은닉화

WhalesJin committed:
docs: 클래스 다이어그램 수정

wonbi92 commented on Calculator/Calculator/CalculatorItem.swift:
```diff
@@ -0,0 +1,16 @@
+
+import UIKit
+
+protocol CalculatorItem {
+    
+}
```
일반적으로 비어있는 프로토콜이나 타입, 익스텐션은 한줄로 줄여주는게 좋긴합니다. ㅎㅎ 꼭 이게 맞다는 아니니 참고만 해주세요~
```suggestion
protocol CalculatorItem { }
```

WhalesJin replied:
오!! 사실 개행을 해놓으면서 뭔가 되게 비어있는 느낌이라 많이 허전했는데
저렇게 한 줄로 적어놓으니 완전 깔끔하네요!!
감사합니다 원비 🤩

wonbi92 commented on Calculator/Calculator/CalculatorItem.swift:
```diff
@@ -0,0 +1,16 @@
+//  Created by Whales on 2023/05/31.
+//
+
+import UIKit
```
현재 소스파일에서 `UIKit`이 필요할까요? 🤔 `UIKit`을 인포트 하신 이유가 궁금합니다!

WhalesJin replied:
미션을 진행하면서 결국에는 UI요소들과 연결해야할거라는 생각에 `UIKit`을 임포트 했습니다!!
근데 지금 생각해보니 이 친구는 프로토콜을 정의하는 부분이니 필요없을거라고 판단되네요.
수정하겠습니다 감사합니다 🫢

wonbi92 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,36 @@
+//  Created by Whales on 2023/05/31.
+//
+
+import UIKit
```
요기도 마찬가지네요! 현재 소스파일에서 `UIKit`이 필요할까요? 🤔 `UIKit`을 인포트 하신 이유가 궁금합니다!

WhalesJin replied:
미션을 진행하면서 결국에는 UI요소들과 연결해야할거라는 생각에 `UIKit`을 임포트 했습니다!!

wonbi92 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,36 @@
+
+import UIKit
+
+struct CalculatorItemQueue<CalculatorItem> {
```
제네릭을 활용하셨군요! 👍 제네릭을 활용하면 어떤 장점이 있을까요?

추가적으로, 제네릭의 타입명이 프로토콜과 같아서 코드를 읽는 사람입장에서는 헷갈릴 수도 있을거 같군요..! 더 좋은 네이밍을 고려해볼까요?

그런데 현재 로직대로라면 "`CalculatorItemQueue`에서 다루는 요소는 `CalculateItem` 프로토콜을 준수합니다." 라는 요구조건을 만족할 수 없을거 같아요..! 제네릭으로 받아오는 타입에 제한을 줘야할거 같은데 어떻게 해야 할까요? 🤔

WhalesJin replied:
제네릭을 사용하면 특정 타입에 얽매이지 않고 코드를 구현 할 수 있어서 유연하고 재사용 가능한 함수와 타입을 작성할 수 있다고 읽었습니다. 이번 프로젝트 같은 경우에는 `queue`에 연산기호와 숫자가 들어갈거라 `Double`과 `String`타입이 `CalculateItem` 프로토콜을 준수하도록 `extension`해줬습니다!!
그리고 제네릭 타입명으로는 `Element`를 넣어주었습니다!!

wonbi92 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,36 @@
+
+import UIKit
+
+struct CalculatorItemQueue<CalculatorItem> {
+    private var calculatorItemQueue: [CalculatorItem] = []
+    
+    var count: Int {
+        return calculatorItemQueue.count
+    }
+    
+    var isEmpty: Bool {
+        return calculatorItemQueue.isEmpty
+    }
+    
+    mutating func enqueue(_ element: CalculatorItem) {
+        calculatorItemQueue.append(element)
+    }
+    
+    mutating func dequeue() {
+        calculatorItemQueue.removeFirst()
+    }
+    
+    mutating func clearQueue() {
+        calculatorItemQueue.removeAll()
+    }
+    
+    func readCalculatorItemQueue() -> [CalculatorItem] {
+        return calculatorItemQueue
+    }
+}
```
현재 `Queue` 구조체가 굳이 필요한지에 대한 의문이 드는군요..! 🤔 각각의 메서드와 연산프로퍼티 모두 추가적인 계산이나 로직 없이 배열에 있는 프로퍼티와 메서드를 그대로 사용하고 있습니다. 이런 식이라면.. 굳이 `CalculatorItemQueue`라는 타입 없이 바로 배열을 사용해도 상관없을것 같네요..!

WhalesJin replied:
곰곰히 생각해보고 더 공부해서 코드 수정했습니다!!
답변이 길어져 저기 아래쪽에 코멘트 작성했습니다 🙇🏻‍♀️

wonbi92 commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,65 @@
+    }
+    
+    func test_enqueue함수호출시_배열의원소개수가_늘어나는지() {
+        // given
+        let input = 3
+        let expectation = 1
+        // when
+        sut.enqueue(input)
+        // then
+        XCTAssertEqual(sut.count, expectation)
```
given, when, then 👍👍

wonbi92 commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,65 @@
+    }
+    
+    func test_enqueue함수호출시_배열의원소가_늘어나는지() {
+        // given
+        sut.enqueue(2)
+        let input = 3
+        let expectation = [2, 3]
+        // when
+        sut.enqueue(input)
+        // then
```
로직 자체가 간단하기 떄문에 굳이 요런식으로 작성할 필요 없이 바로 데이터를 입력해주는게 조금 더 읽기 쉬워보이네요.
그리고, 2는 바로 인큐하고, 3은 상수로 할당한 후에 인큐하는 방식은 컨벤션의 통일성이 없어서 가독성이 매우 떨어지는것 같습니다. 🥲
```suggestion
        // given
        let expectation = [2, 3]
        // when
        sut.enqueue(2)
        sut.enqueue(3)
        // then
```

WhalesJin replied:
테스트 코드를 짤 때 배열의 은닉화를 풀고 배열 = [2] 로 직접 넣어준 뒤, 3을 넣었을 때 앞으로 들어가는게 아니라 뒤로 들어간다는 걸 눈으로 확인하고 싶어서 만들었던 코드였습니다.
근데 은닉화를 다시 해주면서 테스트 코드에 나는 오류부분을 고치다보니 저기 한 줄만 후루룩 고쳐버렸네요ㅠㅠ
앞으로 코드를 짜거나 수정을 할 때 이런 부분들을 좀 더 신경 써 보겠습니다!!
감사합니다 🙇🏻‍♀️

wonbi92 commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,65 @@
+        XCTAssertEqual(sut.readCalculatorItemQueue(), expectation)
+    }
+    
+    func test_clearQueue함수호출시_배열의원소가_다삭제되면_true를_반환한다() {
```
테스트 케이스를 작성하는 방법은 사바사, 회바회기 때문에 정답은 없습니다만, 각 테스트 케이스별 표현 방식은 통일 시켜주시는게 좋겠죠? 
이전 테스트 케이스들은 `~~ 하는지`로 작성하셨으니 이 테스트 케이스 역시 동일하게 작성해주는게 좋습니다!
```suggestion
    func test_clearQueue함수호출시_배열의원소가_다삭제되면_true를_반환하는지() {
```

WhalesJin replied:
오!! 그런 부분도 통일 시켜주면 좋군요!!
넵넵 수정하고 앞으로도 신경 쓰도록 해보겠습니다!!! 감사합니다 😆

wonbi92 commented on Calculator/Calculator/CalculatorItem.swift:
```diff
@@ -0,0 +1,16 @@
+
+import UIKit
+
+protocol CalculatorItem {
```
요구사항의 프로토콜 명은 `CalculateItem`이네요! 요구사항을 정확하게 파악하는것! 매우 중요하겠죠? 😁

WhalesJin replied:
🫢!!!!!
이런 실수를...!!
수정하고 앞으로 더 신경쓰겠습니다!!!

wonbi92 reviewed (changes requested):
@WhalesJin 
반갑습니다 웨일! 3주동안 계산기 리뷰를 맡게된 Wonbi입니다! 앞으로 잘 부탁드려요~ 😁

step1진행하시느라 고생하셨습니다!
고민하신 흔적이 PR메시지와 코드에서 잘 나타나서 좋았습니다! 👍

몇가지 수정사항과 참고하실만한 부분들을 코맨트 드렸으니 쭉 읽어보시고 궁금하신 점 있으시면 말씀해주세요~ 💪

### 아래는 질문주신 부분에 대한 제 생각입니다! 😁
---

> AC버튼을 누르면 초기화 되어야해서 clearQueue()라는 메서드를 만들었는데,
인스턴스 자체가 초기화가 된다면 큐 타입 안에 따로 메서드를 만들 필요가 없지 않을까 라는 생각이 들었습니다.
두 가지 생각에서 고민하다가 메서드를 만들었는데 인스턴스 자체를 초기화 하는 쪽과 메서드로 큐 안에 있는 데이터만 초기화 하는 쪽 두 가지 중 어떤 게 더 좋은 방법일지 조언을 얻고 싶습니다.
- 흠.. 글쎄요 사실 두 방법이 로직적으로 큰 차이는 없어보이긴 합니다. 결국 배열의 `removeAll()`메서드를 호출하는가 새로운 큐를 만드는가의 차이일거 같네요!
- 저는 객체 하나를 통으로 들어내고 새로 넣는것보다는 내부의 배열만 초기화 하는 것이 아주 미약하지만.. 아주 쪼금의 성능적 이득이 있다고 생각하긴 합니다. 다만 정말정말 미약하기 때문에.. 거의 차이가 없다고 생각해도 될거 같아요 ㅎㅎ;

> test를 먼저 만들면 함수가 아직 만들어지지 않은 상태이기에 에러가 나있는 상태로 커밋을 하게 됩니다. 에러가 있는데 commit을 한다는게 너무 이상한데 TDD의 과정을 남기기 위해서 일단 그 상태로 commit을 진행했습니다.
이 commit 방법이 맞는지 더 좋은 다른 방법이 있는지 조언 부탁드립니다.🙏🏻
- TDD를 처음하면 너무 어색하고 시간도 오래걸리고 어렵죠 ㅎㅎ;
- TDD는 개방 방향성에 대한 방법론이기 때문에 커밋의 방법이 맞다 틀리다 할 수는 없습니다. 웨일이 생각하기에 이 방법과 기준이 적절하겠다 라고 판단하시는거죠!
- 틀린 테스트를 커밋하는 것이 너무 어색하다면, 틀린 테스트를 해결하는 것까지 하나의 커밋 단위로 두는것도 좋은 방법일 수 있겠네요!

> 또 Unit test를 진행할 때는 테스트 함수 명을 test_함수이름으로 시작하게 짓는다고 들었는데, 그러면 함수 이름을 먼저 만들어두면 좋은지, 아니면 테스트 만들고commit 후에 함수 만들면서 테스트 함수 이름도 수정해야하는지 혼란스럽습니다. 좋은 조언이 있으시다면 부탁드립니다 🙏🏻
- 테스트 메서드명은 테스트를 작성할 때 같이 작성하는 편이긴 합니다. 하지만 모든 테스트 메서드명을 작성하실 필요는 없고, 하나의 테스트 케이스를 작성할 때마다 새로운 메서드명을 짓는게 편하긴 하겠죠?
- 추가적으로 저는, TDD를 할 때 앱이 동작하기 위해 필요한 객체들이 어떤 행동을 해야 하는지를 먼저 정의한 후에, 그 행동을 수행하기 위한 작은 행동들을 정의하고, 이를 하나의 작은 행동들로 쪼개질 때 까지 반복하여 정리하는 작업을 먼저 수행합니다. 그 다음 각 행동에 대한 테스트 케이스를 작성하는 과정을 시작하죠.
- 요런 방식으로 TDD를 수행한다면, 웨일이 말씀하신대로 빈 메서드를 미리 다 작성해놓을 수도 있겠네요 ㅎㅎ

### 아래는 고민하신 부분에 대한 제 의견입니다. 😉
---
> 단방향 연결리스트의 가장 장점은 리스트 중간에서의 추가 / 삭제가 자유롭다는 부분인데 계산기 프로젝트에서는 중간 수정이 없어서 단방향 연결리스트보다는 크기가 정해져있고 새로운 요소 삽입이 빠른 배열이 더 효율적이라고 생각해서 배열로 선택하였습니다.

- 링크드리스트가 정말 중간에서의 추가 삭제가 자유로울까요? 중간에 있는 특정한 위치에 값을 추가하거나 그 값을 삭제하기 위해 그 위치까지 찾아가는 동작이 필요할거 같은데요 🤔
- 현재 큐를 구현하시기 위해 배열을 선택하신 근거가 조금 빈약한거 같습니다. 자료구조 Queue와 Stack의 차이는 무엇일까요? Array를 통한 Queue 구현의 단점은 무엇이 있을까요?

> 작성해주신 UML에 대한 코맨트 입니다.
- 현재 UML상에서 `CalculatorItemQueue`가 `CalculatorItem`프로토콜을 의존하고 있다고 작성하셨는데, 코드상으로는 그렇지 않은거 같아요..! 다시한번 요구사항을 정확히 파악하시고 코드를 수정해볼까요?

WhalesJin committed:
refactor: 네이밍과 컨벤션 수정

WhalesJin committed:
README.md 작성

WhalesJin committed:
feat: 연결 리스트 구현을 위한 Node 클래스 생성

WhalesJin committed:
feat: 연결 리스트 클래스 생성

- head, tail, count, isEmpty 프로퍼티 생성

WhalesJin committed:
feat: 연결 리스트의 테스트 클래스 생성

WhalesJin committed:
test: node를 추가하는 append 함수의 테스트 함수 생성

- node의 개수가 늘어나는지 테스트

WhalesJin committed:
feat: node를 맨 뒤에 추가하는 append 함수 구현

WhalesJin committed:
test: node를 추 가하는 append 함수의 테스트 함수 추가

- isEmpty 상태면 head와 tail에 node가 들어가는지  데이터로 확인
- 데이터 확인을 위한 head와 tail의 data 값을 읽어오는 함수 생성

WhalesJin committed:
refactor: append 함수의 테스트 함수 수정

- isEmpty 상태면 새로운 node에 head, tail 지정

WhalesJin committed:
test: node를 추가하는 append 함수의 테스트 함수 추가

- 전달인자가 맨 뒤로 들어가는지 테스트

WhalesJin committed:
test: 특정한 자리에 node를 추가하는 insert 함수의 테스트 함수 추가

- 추가 시 list의 node가 늘어나는지 테스트

WhalesJin committed:
feat: 특정한 자리에 node를 추가하는 insert 함수 구현

- test 성공

WhalesJin committed:
test: insert 함수의 테스트 함수 추가

- at의 전달인자가 0이면 맨 앞에 node 추가

WhalesJin committed:
refactor: at의 전달인자가 0이면 맨 앞에 node 추가되도록 insert 함수 수정

- test 성공

WhalesJin committed:
test: insert 함수의 테스트 함수 추가

- at의 전달인자가 node 개수 이상이면 추가하지 않는지 테스트

WhalesJin committed:
test: insert 함수의 테스트 함수 추가

- data의 전달인자 값이 at 전달인자 자리의 data로 들어가는지 테스트. test 성공

WhalesJin committed:
test: 첫번째 node를 삭제하는 removeFirst 함수의 테스트 함수 추가

- 삭제시 head가 두번째 node로 옮겨가는지 테스트

WhalesJin committed:
feat: removeFirst 함수 구현

- test 성공

WhalesJin committed:
test: 마지막 node를 삭제하는 removeLast 함수의 테스트 함수 추가

- 삭제시 tail이 직전 node로 옮겨가는지 테스트

WhalesJin committed:
feat: removeLast 함수 구현

- test 성공

WhalesJin committed:
test: 특정한 자리의 node를 삭제하는 delete 함수의 테스트 함수 추가

- list의 node 개수가 줄어드는지 테스트

WhalesJin committed:
feat: delete 함수 구현

- test 성공

WhalesJin committed:
test: list의 모든 node를 삭제하는 delete 함수의 테스트 함수 추가

- isEmpty가 true가 되는지 테스트

WhalesJin committed:
feat: delete 함수 구현

- test 성공

WhalesJin committed:
refactor: 연결 리스트 클래스를 구조체로 수정

WhalesJin committed:
refactor: CalculatorItemQueue 구조체의 list 타입을 연결 리스트로 변경

WhalesJin committed:
test: 변경에 따른 테스트 함수 수정

WhalesJin commented:
답변이 길어져서 여기에 남깁니다 🙇🏻‍♀️

### `Queue` 구현을 위한 `list` 타입 선택
처음에는 익숙한 배열로 Queue를 구현했습니다. 구현하는 동안 배열의 메서드들을 쓰면서 큐의 구현의 필요성이 안 느껴지고 뭔가 이상하다라는 생각을 했고, 제대로 배우기 위해서 다시 찾아보고 공부해봤습니다. 배열이랑 연결리스트는 시간복잡도면에서 차이가 있다는 것을 공부했습니다.
- 배열
    - removeFirst : 앞으로 한 칸씩 땡겨져서 O(n)
    - append도 새로운 공간을 할당하고 그 전 데이터를 복사해와야해서 O(n)이 될 수도 있다.
    - 원소 접근 O(1)
- 연결리스트 
    - 요소와 요소를 노드로 연결.
    - append나 removeFirst가 O(1)
    - 단점 : 접근 시간. 포인트를 거쳐가면서 접근하기 때문에 오래 걸린다.

이러한 차이점들을 보고 계산기 프로젝트는 큐로 구현하기 때문에 중간 삽입이나 삭제가 없어서 배열의 장점이자 연결리스트의 단점인 접근면에서 문제가 되지 않고, `dequeue` 메서드를 구현할 때 쓰게 될 `removeFirst` 부분에서 시간복잡도가 낮은 연결리스트가 더 효율적일 것 같아서 연결리스트로 코드를 수정하였습니다.
공부에 도움이 될 것 같아서 처음에는 `양방향 연결 리스트(Doubly Linked List)`로 코드를 구현해보았는데 계산기에는 `단방향 연결 리스트(Singly Linked List)`가 적당할 것 같아서 단방향 연결 리스트로 코드를 작성하였습니다.
<details>
<summary>양방향 연결 리스트</summary>

```swift
class Node<T> {
    var data: T?
    var next: Node?
    var prev: Node?
    
    init(data: T? = nil, next: Node? = nil, prev: Node? = nil) {
        self.data = data
        self.next = next
        self.prev = prev
    }
}

class DoublyLinkedList<T> {
    private var head: Node<T>?
    private var tail: Node<T>?
    
    func append(data: T?) { // 맨 뒤에 추가
        if tail == nil {
            head = Node(data: data)
            tail = head
            return
        }
        // tail의 next가 새 data, tail이 새 data의 prev가 되고, 새 data가 tail이 된다.
        var newNode = Node(data: data)
        tail?.next = newNode
        newNode.prev = tail
        tail = newNode
    }
    
    func insert(data: T?, at index: Int) { // index==0이면 head에 추가, 중간에 추가, index가 너무 크면 tail에
        if tail == nil {
            head = Node(data: data)
            tail = head
            return
        }
        
        if index < 0 {
            return
        } else if index == 0 {
            var newNode = Node(data: data)
            head?.prev = newNode
            newNode.next = head
            head = newNode
            return
        }
        
        // 2번에 넣는다고 치면, 1번의 next가 new, 2번의 Prev가 new, new의 prev가 1번, new의 next가 2번
        var node = head
        for _ in 0..<(index - 1) {
            if node?.next == nil { break }
            node = node?.next
        }
        
        let nextNode = node?.next
        let newNode = Node(data: data)
        node?.next = newNode
        node?.next?.next = nextNode
        nextNode?.prev = newNode
        newNode.prev = node
    }
    
    func removeFirst() {
        if head == nil {
            return
        }
        
        head = head?.next
        head?.prev = nil
    }
    
    func removeLast() {
        if head == nil {
            return
        }
        
        tail = tail?.prev
        tail?.next = nil
    }
    
    func delete(at index: Int) {
        if tail == nil {
            return
        }
        
        if index < 0 {
            return
        } else if index == 0 {
            head = head?.next
            head?.prev = nil
            return
        }
        
        var node = head
        for _ in 0..<(index - 1) {
            if node?.next == nil { return }
            node = node?.next
        }
        
        var temp = node?.next
        node?.next = node?.next?.next
        node?.next?.next?.prev = node
        temp = nil
    }
}
```
</details>

부족한 부분과 수정할 부분이 있다면 조언 부탁드립니다.🙏🏻

WhalesJin committed:
refactor: 연결 리스트 구조체의 count 프로퍼티 은닉화

WhalesJin committed:
feat: 큐를 구현하는 프로토콜 생성

큐에 필요한 프로퍼티와 메서드를 생성

WhalesJin commented:
기존 코드는 큐를 구현할 때마다 같은 맥락의 프로퍼티와 메서드를 구현해야하는 코드를 가지고 있었습니다.
그러던 중 `프로토콜`을 활용하면 좋다는 얘기를 듣고 공부해보니
꼭 들어가야하는 요소만 골라 `추상화`하여 큐를 구현할 수 있는 프로토콜을 만들면
비슷하지만 조금 다른 여러 형태의 큐를 만들기에 좋은 방향이 될 것 같아서
프로토콜을 구현해보았습니다!!

wonbi92 commented on Calculator/Calculator/LinkedList.swift:
```diff
@@ -0,0 +1,146 @@
+//  Created by Whales on 2023/06/03.
+//
+
+class Node<T> {
```
더이상 Node를 상속하는 객체가 없는거 같네요! 상속하는 객체가 없는 클래스는 어떻게 해야 할까요~?

WhalesJin replied:
`final` 키워드를 붙여서 상속을 막고, 상속이 되지 않음을 명시해주면 좋을 것 같아요!!
공부한 부분도 같이 적어보았습니다!!

> `final`
- 문법적 의미
    - `final class` -> 최종 클래스!!
        다른 `class`에서 이 `class`를 상속 받을 수 없다.
    - `final func` 처럼 프로퍼티, 메소드 등에 붙으면
        `override(재정의)`가 불가능해진다.
- 성능적 이점
    - swift의 클래스는 일반적으로 `Dynamic Dispatch` 방식을 이용해 런타임시 어떤 `method`를 호출할 지 결정한다.
        - `Dynamic Dispatch` : 런타임에 호출될 함수를 결정하는 방식
            함수 호출과정에서 실제 참조할 요소를 찾는 과정이 있어 `Static Dispatch` 방식에 비해 속도가 느리다. 
        - `Static Dispatch` : 컴파일타임에 실제 호출할 함수를 결정하는 방식
            함수 호출이 간단하고 속도가 빠르다.
    - `final` 키워드를 붙여서 재정의가 불가능해지면 `Dynamic Dispatch` 과정이 필요없음이 명시되어서 `Static Dispatch`를 사용해 컴파일 타임에 호출할 `method`를 지정함으로 런타임 성능이 향상 된다.

wonbi92 commented on Calculator/Calculator/LinkedList.swift:
```diff
@@ -0,0 +1,146 @@
+//
+
+class Node<T> {
+    var data: T?
```
data타입이 꼭 옵셔널이어야 하는가? 에 대한 의문이 드네요..! 일단 노드가 생성될때는 데이터가 존재해야 할거 같은데 말이죠. 🤔
옵셔널이라는 것은 존재하지 않을 수 있는 가능성을 의미하는데, 옵셔널을 붙여주신 이유가 궁금합니다!

WhalesJin replied:
노드라는 게 빈 상자로 존재할 수도 있을 것 같다고 생각해서 옵셔널을 넣었던 건데
연결 리스트에 들어가려면 데이터가 존재할 수 밖에 없을 것 같아서 수정했습니다!!
감사합니다 😄

wonbi92 commented on Calculator/Calculator/LinkedList.swift:
```diff
@@ -0,0 +1,146 @@
+    }
+    
+    mutating func append(data: T?) {
+        if isEmpty {
+            head = Node(data: data)
+            tail = head
+            count += 1
+            return
+        }
+        
+        let newNode = Node(data: data)
+        tail?.next = newNode
+        tail = newNode
+        count += 1
```
사바사가 강한 부분이긴 하지만, 저는 if문을 분기점으로 사용할 경우에는 코드의 들여쓰기 계층을 동일하게 해주는 것을 선호하는 편입니다. 
그리고 이렇게 작성하면 count 변수에 +1해주는 동작은 따로 뺄 수 있겠죠?
참고만 하세요~ ㅎㅎ
```suggestion
        let newNode = Node(data: data)
        
        if isEmpty {
            head = newNode
            tail = head
        } else {
            tail?.next = newNode
            tail = newNode
        }
        
        count += 1
```

WhalesJin replied:
오 깔끔하고 좋은 것 같아요!!!!! 감사합니다!!😃

wonbi92 commented on Calculator/Calculator/LinkedList.swift:
```diff
@@ -0,0 +1,146 @@
+    }
+    
+    mutating func insert(data: T?, at index: UInt) {
+        if index >= count {
+            return
+        } else if index == 0 {
```
out of index의 상황을 고려한 로직이군요! 👍👍👍
고민하신 흔적이 보여서 좋습니다. 다만 로직의 특성상 guard를 쓰는게 더 좋아보여요. guard문은 if문과 어떤 차이가 있을까요?
```suggestion
        guard index < count else { return } 
        
        if index == 0 {
```

WhalesJin replied:
`if`는 만약 `~면 ~해라`라는 의미로 쓰이는 반면에
`guard`는 `~가 아니면 끝내라`라는 의미로 쓰이면서 `early exit`이라는 특징을 가집니다.
그런 의미에서 로직의 특성상 `guard`를 쓰는게 더 좋아보인다고 말씀하신걸로 이해를 하고 코드를 수정하였습니다!!

wonbi92 commented on Calculator/Calculator/LinkedList.swift:
```diff
@@ -0,0 +1,146 @@
+    }
+    
+    mutating func removeAll() {
+        if isEmpty {
+            return
+        }
+        
+        while !isEmpty {
+            removeFirst()
+        }
```
링크드 리스트는 다음 데이터를 참조하는 노드들의 연결로 리스트를 구현합니다. 그리고 스위프트는 ARC를 통해 자동으로 참조를 관리해주죠!
따라서, 해당 로직은 알아서 내부적으로 돌아가는 로직을 굳이 코드로 작성하는 느낌이 드네요! ARC가 메모리를 관리하는 방법에 대해 조금 더 공부해볼까요?

WhalesJin replied:
오!!! 연결을 끊고 데이터까지 지워줘야는 줄 알았는데 아니었군요!!!
알려주신 덕분에 ARC에 대해서 공부해봤습니다!!

#### `ARC`(Auto Reference Counting)
- 참조타입은 메모리가 자동으로 힙에 할당된다. 사용하고 난 후에는 반드시 메모리 해제를 해줘야 하는데 그걸 자동으로 해주는게 ARC
- ARC는 클래스 인스턴스가 더이상 필요하지 않을 때 메모리를 자동으로 해제한다.
- 메모리의 참조횟수(RC)를 계산하여 참조횟수가 0이되면 더이상 사용하지 않는 메모리다. -> 해제
- 모든 인스턴스는 자신의 RC 값을 가지고 있다. (인스턴스 생성 시 힙에 같이 저장)
- 참고링크
    - 🍎 [Docs.Swift - ARC](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/automaticreferencecounting/)
    -  <Img src = "https://github.com/WhalesJin/ios-juice-maker-yagom/assets/124643545/ad865807-d3a4-4345-9c55-d7a2c23761d5" width="20"/> [블로그 - ARC](https://babbab2.tistory.com/26)

wonbi92 commented on Calculator/Calculator/QueueConfigurable.swift:
```diff
@@ -0,0 +1,20 @@
+//  Created by Whales on 2023/06/04.
+//
+
+protocol QueueConfigurable {
+    associatedtype CalculateItem
+    
+    var queue: LinkedList<CalculateItem> { get }
+    var count: Int { get }
+    var isEmpty: Bool { get }
+    
+    mutating func enqueue(_ element: CalculateItem)
+    
+    mutating func dequeue()
+    
+    mutating func clearQueue()
+}
```
Queue 추상화! 👍👍👍
네이밍의 경우, 저는 `Queueable`이 좀 더 자연스럽다 생각합니다만, 웨일만의 이유가 있다면 상관없을거 같아요!

WhalesJin replied:
큐를 구성할 수 있는 요소들을 모아놓은 프로토콜이라 구체적으로 네이밍을 해보았습니다!!

wonbi92 commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,77 @@
+@testable import Calculator
+
+final class CalculatorTests: XCTestCase {
+    var sut: CalculatorItemQueue<Double>!
```
Queue를 추상화 해보셨으니, 좀 더 깔끔한 테스트 방법을 알려드릴게요! Test Double을 이용해서 테스트용 객체를 따로 만들어 테스트를 진행하시는 것도 고려해보세요!
테스트에서도 DIP를 적용할 수 있게 되면서 SOLID원칙도 잘 지킬 수 있게 된답니다! 😉

WhalesJin replied:
알려주신 `Test Double`에 대해 공부해보았습니다!! 생각보다 많은 종류가 있더라고요! 이 중에 현재 저에게 필요하다고 생각이 되는 `Mock` 위주로 일단 공부해서 정리해봤고, 다른 종류들도 차차 공부해 볼 예정입니다!!
감사합니다 원비 =)

---

#### `Test Double`
- `Unit Test`를 작성하는 동안 production에 사용될 객체와 동일하게 동작하지만 단순화된 버전이 필요한 경우에 쓰는 객체.
- 외부 의존성 관리
- 다양한 시나리오를 보다 쉽게 시뮬레이션 가능
- 종류
    - Dummy(실제 사용되지는 않지만 필요한 객체),
    - Fake(실제 구현이 존재, 실환경에서는 사용할 수 없지만 테스트를 위해 사용),
    - Stub(하드코딩된 정해진 결과값을 반환하는 객체),
    - Spy(Stub의 역할을 가지면서 호출된 내용에 대해 약간의 정보를 기록),
    - Mock(행위를 검증하기 위한 객체. 로직이 정상적으로 실행됐는지 여부 판단) 

이 중 현재 상황에 해당하는 Mock만 공부
- `Mock`
    - Mock 객체는 행위를 기반으로 테스트 케이스를 작성
        (일반적인 Test Double은 상태를 기반으로 테스트 케이스를 작성)
    - 호출에 대한 기대를 명세하고 내용에 따라 동작하도록 프로그래밍된 객체
    - 사용하는 경우
        1. 테스트 작성을 위한 환경 구축이 어려울 때
            - 환경 구축을 위한 작업 시간이 많이 필요한 경우 Mock 객체를 사용
            - 특정 모듈을 아직 갖고 있지 않아서 테스트 환경을 구축하지 못할 경우
            - 타 부서와의 협의나 정책이 필요한 경우에도 Mock이 필요.
            - 연계 모듈이라서 다른 쪽에서 승인을 해줘야 테스트가 가능한 경우, 방화벽으로 막혀 있어서 통과가 어려운 경우 등이 이에 속함.
        2. 테스트가 특정 경우나 순간에 의존적일 때
        3. 테스트 시간이 오래 걸리는 경우

    𖤐 이 중 1번 경우에 해당.
    
- 참고링크
    - <Img src = "https://github.com/WhalesJin/ios-juice-maker-yagom/assets/124643545/ad865807-d3a4-4345-9c55-d7a2c23761d5" width="20"/> [블로그 - Test Double(간략)](https://cheonsong.tistory.com/24)
    - <Img src = "https://github.com/WhalesJin/ios-juice-maker-yagom/assets/124643545/ad865807-d3a4-4345-9c55-d7a2c23761d5" width="20"/> [블로그 - Test Double(자세히)](https://velog.io/@leeyoungwoozz/Test-Doubles)

wonbi92 reviewed (approved):
@WhalesJin 
고민하신 흔적들이 잘 보여서 정말 좋았습니다! 👍 고생 많으셨어요!
코맨트로 약간의 수정할 내용들을 말씀 드렸지만, 그 부분은 다음 스텝에서 진행하셔도 괜찮을거라 판단됩니다.
머지해드리겠습니다! 다음 스탭 진행해주세요~
